components/higher-order/with-dragging/index.js

```diff
@@ -0,0 +1,163 @@
+		 * @param  {number} order     The current index of the element to be dragged.
+		 * @param  {string} type      A unique constant identifying the type of dragging operation.
+		 */
+		onDragStart( event, uid, elementId, order, type ) {
+			const element = document.getElementById( elementId );
+			const elementWrapper = element.parentNode;
+			const elementRect = element.getBoundingClientRect();
+			const elementTopOffset = parseInt( elementRect.top, 10 );
+			const elementLeftOffset = parseInt( elementRect.left, 10 );
+			const clone = element.cloneNode( true );
+			const cloneWrapper = document.createElement( 'div' );
+			const cloneNodeId = `clone-wrapper-${ element.id }`;
+			const dragImage = document.createElement( 'div' );
+
+			this.setState( { cloneNodeId, elementId } );
+
+			// Set a fake drag image to avoid browser defaults. Remove from DOM right after.
+			if ( 'function' === typeof event.dataTransfer.setDragImage ) {
+				dragImage.id = `drag-image-${ element.id }`;
+				dragImage.classList.add( this.config.dragImageClass );
+				document.body.appendChild( dragImage );
+				event.dataTransfer.setDragImage( dragImage, 0, 0 );
+				setTimeout( ( ( _dragImage ) => () => {
+					document.body.removeChild( _dragImage );
+				} )( dragImage ), 0 );
+			}
+
+			event.dataTransfer.setData(
+				'text',
+				JSON.stringify( {
+					uid: uid,
+					fromIndex: order,
+					type: type
+				} )
+			);
+
+			// Prepare element clone and append to element wrapper.
+			clone.id = `clone-${ element.id }`;
+			cloneWrapper.id = cloneNodeId;
+			cloneWrapper.classList.add( this.config.cloneWrapperClass );
+			cloneWrapper.style.width = `${ elementRect.width + ( this.config.clonePadding * 2 ) }px`;
+
+			if ( elementRect.height > this.config.cloneHeightTransformationBreakpoint ) {
+				// Scale down clone if original element is larger than 700px.
+				cloneWrapper.style.transform = 'scale(0.5)';
+				cloneWrapper.style.transformOrigin = 'top left';
+				// Position clone near the cursor.
+				cloneWrapper.style.top = `${ parseInt( event.clientY, 10 ) - 100 }px`;
+				cloneWrapper.style.left = `${ parseInt( event.clientX, 10 ) }px`;
+			} else {
+				// Position clone right over the original element (20px padding).
+				cloneWrapper.style.top = `${ elementTopOffset - this.config.clonePadding }px`;
+				cloneWrapper.style.left = `${ elementLeftOffset - this.config.clonePadding }px`;
+			}
+
+			cloneWrapper.appendChild( clone );
+			elementWrapper.appendChild( cloneWrapper );
+
+			// Mark the current cursor coordinates.
+			this.cursorLeft = event.clientX;
+			this.cursorTop = event.clientY;
+			// Update cursor to 'grabbing', document wide.
+			document.body.classList.add( 'dragging' );
+			document.addEventListener( 'dragover', this.onDragOver );
+			event.stopPropagation();
+		}
+
+		render() {
+			return (
+				<OriginalComponent
+					{ ...this.props }
+					onDragStart={ this.onDragStart }
+					onDragEnd={ this.onDragEnd }
```

Comment:
If `onDragStart` and `onDragEnd` are supposed to be the HTML5 event handlers, maybe we can rely on `findDOMNode` to attach them instead of passing them as props. Nothing guarantees at this point that the `OriginalComponent` will use these props.

Reply:
Maybe renaming them to something like `dragInit` and `dragEnd` might make their intention clearer, cause these are actually closures and not event handlers per se (they accept some properties and return an event handler).

Reply:
My bad, sorry. Totally misleading answer above. When I first wrote these two methods they were closures returning the event listeners. I later updated them back to event listeners, just with additional parameters in the case of `onDragStart`. So what you wrote above @youknowriad is valid, except I don't see (don't know) why it would be a problem for the wrapped component to not use the properties - if you are applying the `withDragging` HOC to something, then you are obviously doing that so that you can use the properties, no?